File: examples/subject.ts
import { Memory } from '@letta-ai/memory-sdk';

async function instanceScopedExample() {
  console.log('== Instance-scoped subject example ==');
  const memory = new Memory({ subjectId: 'user_sarah' });

  await memory.initializeMemory('preferences', 'Known user preferences.', 'Likes cats', 10000, true);

  const run = await memory.addMessages([
    { role: 'user', content: 'I love cats' },
  ]);
  await memory.waitForRun(run);

  const raw = await memory.getMemory('preferences');
  console.log('Raw:', raw);

  const formatted = await memory.getMemory('preferences', true);
  console.log('Formatted:', formatted);

  const blocks = await memory.listBlocks();
  console.log('Blocks:', blocks.map((b: any) => b.label));

  await memory.deleteBlock('preferences');
  console.log("Deleted 'preferences'");
}

async function explicitContextExample() {
  console.log('== Explicit subject example ==');
  const memory = new Memory();

  await memory.initializeSubject('project_alpha', true);
  await memory.initializeMemory('spec', 'Project spec', 'v1', 10000, false, 'project_alpha');

  const run = await memory.addMessagesToSubject('project_alpha', [
    { role: 'user', content: 'Kickoff complete' },
  ]);
  await memory.waitForRun(run);

  const spec = await memory.getMemory('spec', false, 'project_alpha');
  console.log('Spec:', spec);
}

async function main() {
  await instanceScopedExample();
  await explicitContextExample();
}

if (require.main === module) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

File: examples/subject.py
from ai_memory_sdk import Memory


def instance_scoped_example():
    print("== Instance-scoped subject example ==")
    memory = Memory(subject_id="user_sarah")

    # Create/ensure a block in this subject
    memory.initialize_memory(
        label="preferences",
        description="Known user preferences.",
        value="Likes cats",
        reset=True,
    )

    # Add messages to the bound subject (unified API)
    run = memory.add_messages([
        {"role": "user", "content": "I love cats"}
    ])
    memory.wait_for_run(run)

    # Read blocks (raw and formatted)
    raw = memory.get_memory("preferences")
    formatted = memory.get_memory("preferences", prompt_formatted=True)
    print("Raw:", raw)
    print("Formatted:", formatted)

    # List and delete
    blocks = memory.list_blocks()
    print("Blocks:", [b.label if hasattr(b, 'label') else b.get('label') for b in blocks])
    memory.delete_block("preferences")
    print("Deleted 'preferences'")


def explicit_subject_example():
    print("== Explicit subject example ==")
    memory = Memory()

    # Initialize a named subject (agent)
    memory.initialize_subject("project_alpha", reset=True)

    # Create a block under this subject
    memory.initialize_memory(
        label="spec",
        description="Project spec",
        value="v1",
        subject_id="project_alpha",
    )

    # Add messages to that subject
    run = memory.add_messages_for_subject("project_alpha", [
        {"role": "user", "content": "Kickoff complete"}
    ])
    memory.wait_for_run(run)

    spec = memory.get_memory("spec", subject_id="project_alpha")
    print("Spec:", spec)


if __name__ == "__main__":
    instance_scoped_example()
    explicit_subject_example()
